File: app/src/main/java/com/example/msdc/activities/SearchActivity.java
package com.example.msdc.activities;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.text.HtmlCompat;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import com.example.msdc.adapter.MovieSearchAdapter;
import com.example.msdc.adapter.TVSearchAdapter;
import com.example.msdc.api.ApiClient;
import com.example.msdc.api.ApiService;
import com.example.msdc.api.MovieResponse;
import com.example.msdc.api.MovieResult;
import com.example.msdc.api.TVResponse;
import com.example.msdc.api.TVResult;
import com.example.msdc.databinding.ActivitySearchBinding;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SearchActivity extends AppCompatActivity {
    private final List<MovieResult> movieResults = new ArrayList<>();
    private final List<TVResult> tvResults = new ArrayList<>();
    private String query;
    private MovieSearchAdapter movieSearchAdapter;
    private TVSearchAdapter tvSearchAdapter;
    private ApiService apiService;
    private int page = 1;
    private ActivitySearchBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivitySearchBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        Toolbar toolbar = binding.toolbar;

        binding.rvSearch.setLayoutManager(new StaggeredGridLayoutManager(3, RecyclerView.VERTICAL));
        binding.loadingSearch.setVisibility(View.VISIBLE);

        movieSearchAdapter = new MovieSearchAdapter(movieResults, this);
        tvSearchAdapter = new TVSearchAdapter(tvResults, this);

        apiService = ApiClient.getClient().create(ApiService.class);
        query = getIntent().getStringExtra("searchFor");
        toolbar.setTitle(HtmlCompat.fromHtml("You searched for : <b>" + query + "</b>",
                HtmlCompat.FROM_HTML_MODE_LEGACY));
        String type = getIntent().getStringExtra("type");

        switch(type){
            case "Movies" :
                searchForMovies(page);
                binding.rvSearch.setAdapter(movieSearchAdapter);

                binding.rvSearch.addOnScrollListener(new RecyclerView.OnScrollListener() {
                    @Override
                    public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                        super.onScrolled(recyclerView, dx, dy);

                        if(!binding.rvSearch.canScrollVertically(1)){
                            page++;
                            searchForMovies(page);
                        }
                    }
                });
                break;

            case "TV Shows" :
                searchForTV(page);
                binding.rvSearch.setAdapter(tvSearchAdapter);

                binding.rvSearch.addOnScrollListener(new RecyclerView.OnScrollListener() {
                    @Override
                    public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                        super.onScrolled(recyclerView, dx, dy);

                        if(!binding.rvSearch.canScrollVertically(1)){
                            page++;
                            searchForTV(page);
                        }
                    }
                });
                break;
        }

        toolbar.setOnClickListener(v-> onBackPressed());
    }

    private void searchForMovies(int page) {
        int limit = 15;

        Call<MovieResponse> call = apiService.searchMovie(MainActivity.MY_API_KEY, MainActivity.LANGUAGE, query, page, limit);
        call.enqueue(new Callback<MovieResponse>() {
            @Override
            public void onResponse(@NonNull Call<MovieResponse> call, @NonNull Response<MovieResponse> response) {
                if(response.body() != null){
                    if(!response.body().getResult().isEmpty()){
                        binding.loadingSearch.setVisibility(View.GONE);

                        int oldCount = movieResults.size();
                        movieResults.addAll(response.body().getResult());
                        movieSearchAdapter.notifyItemRangeInserted(oldCount, movieResults.size());
                    } else {
                        binding.loadingSearch.setVisibility(View.GONE);
                        binding.textNoResults.setVisibility(View.VISIBLE);
                    }
                }
            }

            @SuppressLint("SetTextI18n")
            @Override
            public void onFailure(@NonNull Call<MovieResponse> call, @NonNull Throwable t) {
                binding.loadingSearch.setVisibility(View.GONE);

                binding.textNoResults.setText("No Results, Try Search Again !!!");
                binding.textNoResults.setVisibility(View.VISIBLE);
            }
        });
    }

    private void searchForTV(int page) {
        int limit = 15;

        Call<TVResponse> call = apiService.searchTv(MainActivity.MY_API_KEY, MainActivity.LANGUAGE, query, page, limit);
        call.enqueue(new Callback<TVResponse>() {
            @Override
            public void onResponse(@NonNull Call<TVResponse> call, @NonNull Response<TVResponse> response) {
                if(response.body() != null){
                    if(!response.body().getResult().isEmpty()){
                        binding.loadingSearch.setVisibility(View.GONE);

                        int oldCount = tvResults.size();
                        tvResults.addAll(response.body().getResult());
                        tvSearchAdapter.notifyItemRangeInserted(oldCount, tvResults.size());
                    } else {
                        binding.loadingSearch.setVisibility(View.GONE);
                        binding.textNoResults.setVisibility(View.VISIBLE);
                    }
                }
            }

            @SuppressLint("SetTextI18n")
            @Override
            public void onFailure(@NonNull Call<TVResponse> call, @NonNull Throwable t) {
                binding.loadingSearch.setVisibility(View.GONE);

                binding.textNoResults.setText("No Results, Try Search Again !!!");
                binding.textNoResults.setVisibility(View.VISIBLE);
            }
        });
    }
}

File: app/src/main/java/com/example/msdc/activities/SearchActivity.kt
package com.example.msdc.activities

import android.content.ContentValues
import android.content.res.Configuration
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.msdc.adapter.MovieVerticalAdapter
import com.example.msdc.adapter.TVVerticalAdapter
import com.example.msdc.api.ApiClient
import com.example.msdc.api.ApiService
import com.example.msdc.api.MovieResponse
import com.example.msdc.api.MovieResult
import com.example.msdc.api.TVResponse
import com.example.msdc.api.TVResult
import com.example.msdc.databinding.ActivitySearchBinding
import com.google.firebase.crashlytics.FirebaseCrashlytics
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class SearchActivity : AppCompatActivity() {
    private var apiService: ApiService? = null

    private var movieSearchAdapter: MovieVerticalAdapter? = null
    private var tvSearchAdapter: TVVerticalAdapter? = null

    private val movieResults: MutableList<MovieResult> = ArrayList()
    private val tvResults: MutableList<TVResult> = ArrayList()

    private var page = 1
    private val limit = 15
    private var query: String? = null

    private lateinit var searchToolbar: Toolbar
    private lateinit var loadingSearch: ProgressBar
    private lateinit var rvSearch: RecyclerView
    private lateinit var textNoResults: TextView
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(true)

        val binding = ActivitySearchBinding.inflate(layoutInflater)
        setContentView(binding.root)

        searchToolbar = binding.toolbar
        loadingSearch = binding.loadingSearch
        rvSearch = binding.rvSearch
        textNoResults = binding.textNoResults

        val mPortraitLayoutManager = GridLayoutManager(this, 3)
        val mLandscapeLayoutManager = GridLayoutManager(this, 5)

        if(applicationContext.resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE){
            rvSearch.layoutManager = mLandscapeLayoutManager
        } else {
            rvSearch.layoutManager = mPortraitLayoutManager
        }

        movieSearchAdapter = MovieVerticalAdapter(movieResults, this)
        tvSearchAdapter = TVVerticalAdapter(tvResults, this)
        apiService = ApiClient.client!!.create(ApiService::class.java)
        query = intent.getStringExtra("searchFor")

        searchToolbar.title = "Search Results"
        searchToolbar.subtitle = query

        when (intent.getStringExtra("type")) {
            "Movies" -> {
                searchForMovies(page)
                binding.rvSearch.adapter = movieSearchAdapter
                binding.rvSearch.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                        super.onScrolled(recyclerView, dx, dy)
                        if (!binding.rvSearch.canScrollVertically(1)) {
                            page++
                            searchForMovies(page)
                        }
                    }
                })
            }

            "TV Shows" -> {
                searchForTV(page)
                binding.rvSearch.adapter = tvSearchAdapter
                binding.rvSearch.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                        super.onScrolled(recyclerView, dx, dy)
                        if (!binding.rvSearch.canScrollVertically(1)) {
                            page++
                            searchForTV(page)
                        }
                    }
                })
            }
        }

        onBackPressedDispatcher.addCallback(
            this,
            object: OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    Log.d(ContentValues.TAG, "Activity back pressed invoked")
                    finish()
                }
            }
        )

        searchToolbar.setOnClickListener {
            onBackPressedDispatcher.onBackPressed()
        }
    }

    private fun searchForMovies(page: Int) {
        val call = apiService?.searchMovie(
            MainActivity.MY_API_KEY,
            MainActivity.LANGUAGE,
            query,
            page,
            limit
        )
        call!!.enqueue(object : Callback<MovieResponse?> {
            override fun onResponse(
                call: Call<MovieResponse?>,
                response: Response<MovieResponse?>
            ) {
                if (response.body() != null) {
                    if (response.body()!!.result!!.isNotEmpty()) {
                        loadingSearch.visibility = View.GONE
                        rvSearch.visibility = View.VISIBLE
                        val oldCount = movieResults.size
                        movieResults.addAll(response.body()!!.result!!)
                        movieSearchAdapter?.notifyItemRangeInserted(oldCount, movieResults.size)
                    } else {
                        loadingSearch.visibility = View.GONE
                        textNoResults.visibility = View.VISIBLE
                    }
                }
            }

            override fun onFailure(call: Call<MovieResponse?>, t: Throwable) {
                loadingSearch.visibility = View.GONE
                Toast.makeText(
                    this@SearchActivity, t.message + " cause : "
                            + t.cause, Toast.LENGTH_SHORT
                ).show()
            }
        })
    }

    private fun searchForTV(page: Int) {
        val call = apiService?.searchTv(
            MainActivity.MY_API_KEY,
            MainActivity.LANGUAGE,
            query,
            page,
            limit
        )
        call!!.enqueue(object : Callback<TVResponse?> {
            override fun onResponse(call: Call<TVResponse?>, response: Response<TVResponse?>) {
                if (response.body() != null) {
                    if (response.body()!!.result!!.isNotEmpty()) {
                        loadingSearch.visibility = View.GONE
                        rvSearch.visibility = View.VISIBLE
                        val oldCount = tvResults.size
                        tvResults.addAll(response.body()!!.result!!)
                        tvSearchAdapter?.notifyItemRangeInserted(oldCount, tvResults.size)
                    } else {
                        loadingSearch.visibility = View.GONE
                        textNoResults.visibility = View.VISIBLE
                    }
                }
            }

            override fun onFailure(call: Call<TVResponse?>, t: Throwable) {
                loadingSearch.visibility = View.GONE
                Toast.makeText(
                    this@SearchActivity, t.message + " cause : "
                            + t.cause, Toast.LENGTH_SHORT
                ).show()
            }
        })
    }
}